lib/Transforms/LLZKLoweringUtils.cpp

```diff
@@ -0,0 +1,179 @@
+    Value self = getSelfValueFromCompute(computeFunc);
+    Value rebuilt = builder.create<FieldReadOp>(
+        readOp.getLoc(), readOp.getType(), self, readOp.getFieldNameAttr().getAttr()
+    );
+    return memo[val] = rebuilt;
+  }
+
+  if (auto add = val.getDefiningOp<AddFeltOp>()) {
+    Value lhs = rebuildExprInCompute(add.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(add.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<AddFeltOp>(add.getLoc(), add.getType(), lhs, rhs);
+  }
+
+  if (auto sub = val.getDefiningOp<SubFeltOp>()) {
+    Value lhs = rebuildExprInCompute(sub.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(sub.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<SubFeltOp>(sub.getLoc(), sub.getType(), lhs, rhs);
+  }
+
+  if (auto mul = val.getDefiningOp<MulFeltOp>()) {
+    Value lhs = rebuildExprInCompute(mul.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(mul.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<MulFeltOp>(mul.getLoc(), mul.getType(), lhs, rhs);
+  }
+
+  if (auto neg = val.getDefiningOp<NegFeltOp>()) {
+    Value operand = rebuildExprInCompute(neg.getOperand(), computeFunc, builder, memo);
+    return memo[val] = builder.create<NegFeltOp>(neg.getLoc(), neg.getType(), operand);
+  }
+
+  if (auto div = val.getDefiningOp<DivFeltOp>()) {
+    Value lhs = rebuildExprInCompute(div.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(div.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<DivFeltOp>(div.getLoc(), div.getType(), lhs, rhs);
+  }
+
+  if (auto c = val.getDefiningOp<FeltConstantOp>()) {
+    return memo[val] = builder.create<FeltConstantOp>(c.getLoc(), c.getValue());
+  }
+
+  llvm::errs() << "Unhandled op in rebuildExprInCompute: " << val << '\n';
+  llvm_unreachable("Unsupported op kind");
+}
+
+LogicalResult checkForAuxFieldConflicts(StructDefOp structDef, StringRef prefix) {
+  bool conflictFound = false;
+
+  structDef.walk([&](FieldDefOp fieldDefOp) {
+    if (fieldDefOp.getName().starts_with(prefix)) {
+      (fieldDefOp.emitError() << "Field name '" << fieldDefOp.getName()
+                              << "' conflicts with reserved prefix '" << prefix << '\'')
+          .report();
+      conflictFound = true;
+    }
+  });
+
+  return failure(conflictFound);
+}
+
+void replaceSubsequentUsesWith(Value oldVal, Value newVal, Operation *afterOp) {
+  assert(afterOp && "afterOp must be a valid Operation*");
+
+  for (auto &use : llvm::make_early_inc_range(oldVal.getUses())) {
+    Operation *user = use.getOwner();
+
+    // Skip uses that are:
+    // - Before afterOp in the same block.
+    // - Inside afterOp itself.
+    if ((user->getBlock() == afterOp->getBlock()) &&
+        (user->isBeforeInBlock(afterOp) || user == afterOp)) {
```

Comment:
nit: performance, simple check first
```suggestion
        (user == afterOp || user->isBeforeInBlock(afterOp))) {
```